Sources/Workspace/Workspace.swift

```diff
@@ -4185,6 +4189,219 @@ extension Workspace.Location {
+        ) {
+
+            let sync = DispatchGroup()
+            let transformations = ThreadSafeKeyValueStore<PackageDependency, PackageIdentity>()
+            for dependency in manifest.dependencies {
+                if case .sourceControl(let settings) = dependency, case .remote(let url) = settings.location  {
+                    sync.enter()
+                    self.mapRegistryIdentity(url: url, observabilityScope: observabilityScope, callbackQueue: callbackQueue) { result in
+                        defer { sync.leave() }
+                        switch result {
+                        case .failure(let error):
+                            return completion(.failure(error))
+                        case .success(.some(let identity)):
+                            transformations[dependency] = identity
+                        case .success(.none):
+                            // no identity found
+                            break
+                        }
+                    }
+                }
+            }
+
+            // update the manifest with the transformed dependencies
+            sync.notify(queue: .sharedConcurrent) {
+                do {
+                    let updatedManifest = try self.transformManifest(
+                        manifest: manifest,
+                        transformations: transformations.get(),
+                        observabilityScope: observabilityScope
+                    )
+                    completion(.success(updatedManifest))
+                }
+                catch {
+                    return completion(.failure(error))
+                }
+            }
+        }
+
+        private func transformManifest(
+            manifest: Manifest,
+            transformations: [PackageDependency: PackageIdentity],
+            observabilityScope: ObservabilityScope
+        ) throws -> Manifest {
+            var modifiedDependencies = [PackageDependency]()
+            for dependency in manifest.dependencies {
+                var modifiedDependency = dependency
+                if let registryIdentity = transformations[dependency] {
+                    guard case .sourceControl(let settings) = dependency, case .remote = settings.location else {
+                        // an implementation mistake
+                        throw InternalError("unexpected non-source-control dependency")
+                    }
+                    switch settings.requirement {
+                    case .exact(let version):
+                        modifiedDependency = .registry(identity: registryIdentity, requirement: .exact(version), productFilter: settings.productFilter)
+                    case .range(let range):
+                        modifiedDependency = .registry(identity: registryIdentity, requirement: .range(range), productFilter: settings.productFilter)
+                    case .revision:
+                        // FIXME: warn?
```

Comment:
Deprecation warning maybe? Though feels more like it should be "info" during the initial launch.

Reply:
found a solution in which we only replace the identity instead of full swizzle. works well!